File: Funcional/src/Album.hs
module Album
  ( album,
  )
where

import Util
import Text.Printf

album :: IO ()
album = do
  clearScr
  putStrLn "          ======================================================      "
  putStrLn "                █████╗ ██╗     ██████╗ ██╗   ██╗███╗   ███╗           "
  putStrLn "               ██╔══██╗██║     ██╔══██╗██║   ██║████╗ ████║           "
  putStrLn "               ███████║██║     ██████╔╝██║   ██║██╔████╔██║           "
  putStrLn "               ██╔══██║██║     ██╔══██╗██║   ██║██║╚██╔╝██║           "
  putStrLn "               ██║  ██║███████╗██████╔╝╚██████╔╝██║ ╚═╝ ██║           "
  putStrLn "               ╚═╝  ╚═╝╚══════╝╚═════╝  ╚═════╝ ╚═╝     ╚═╝           "
  putStrLn "          ======================================================      "
  menuAlbum

menuAlbum :: IO ()
menuAlbum = do
  putStrLn "                                                            "
  putStrLn "                           1. ESTATISTICAS                  "
  putStrLn "                           2. MENU INICIAL                  "
  putStrLn "                                                            "
  entradaAlbum

entradaAlbum :: IO()
entradaAlbum = do
  opc <- getLine :: IO String
  navegacaoAlbum opc


navegacaoAlbum :: String -> IO()
navegacaoAlbum opc
  | opc == "1" = estatisticas
  | opc == "2" = putStrLn ""
  | otherwise = do
    putStr "\nDigite uma opção válida\n"
    entradaAlbum


estatisticas :: IO()
estatisticas = do
  putStrLn  "                                                          "
  putStrLn  "---------------  ESTATÍSTICAS ----------------------------"
  conteudo <- lerArquivo "/src/arquivos/album.txt"
  putStrLn  "        ----Quantidade de Figurinhas coladas---        "
  let quantidadeColadas = contaVerdadeiros conteudo
  printf "%.0f \n" quantidadeColadas
  putStrLn "                                                        "
  putStrLn "        ----Quantidade de Figurinhas faltantes---       "
  print(contaFalsos conteudo)
  putStrLn "                                                        "
  putStrLn "        ----Porcentagem figurinhas no álbum (%)---      "
  printf "%.2f \n" ((quantidadeColadas / 250.0)* 100)
  putStrLn ""
  continuar


contaVerdadeiros :: [String] -> Double
contaVerdadeiros [] = 0
contaVerdadeiros (a:as) = do
  if a == "True" then 1.0 + contaVerdadeiros as
  else 0.0 + contaVerdadeiros as

contaFalsos :: [String] -> Int
contaFalsos st = length ([s | s <- st,last (reverseString s) == 'F'])

reverseString :: [Char] -> [Char]
reverseString [] = []
reverseString (x:xs) = reverseString xs ++ [x]

File: Logico/album.pl
:-include('util.pl').

album :-
    shell(clear),
    writeln("          ======================================================      "),
    writeln("                █████╗ ██╗     ██████╗ ██╗   ██╗███╗   ███╗           "),
    writeln("               ██╔══██╗██║     ██╔══██╗██║   ██║████╗ ████║           "),
    writeln("               ███████║██║     ██████╔╝██║   ██║██╔████╔██║           "),
    writeln("               ██╔══██║██║     ██╔══██╗██║   ██║██║╚██╔╝██║           "),
    writeln("               ██║  ██║███████╗██████╔╝╚██████╔╝██║ ╚═╝ ██║           "),
    writeln("               ╚═╝  ╚═╝╚══════╝╚═════╝  ╚═════╝ ╚═╝     ╚═╝           "),
    writeln("          ======================================================      "),
    menuAlbum.


menuAlbum :-
    writeln("                                                            "),
    writeln("                           1. ESTATISTICAS                  "),
    writeln("                           2. MENU INICIAL                  "),
    writeln("                                                            "),
    entradaAlbum.


entradaAlbum:-
    read_line_to_string(user_input, Opc),
    navegacaoAlbum(Opc).


navegacaoAlbum("1"):- estatistica, album.
navegacaoAlbum("2"):- write('').
navegacaoAlbum(_):- write("Digite uma opção válida\n"), entradaAlbum.

estatistica :-
    shell(clear),
    getAlbum(Album),
    contaFigurinhas(Album, Possui, NPossui),
    writeln("                                                        "),
    writeln("                 <<ESTATÍSTICAS>>                       "),
    writeln("                                                        \n"),
    writeln("        <<Quantidade de Figurinhas coladas>>            \n"),
    writeln(Possui),
    writeln("        <<Quantidade de Figurinhas faltantes>>            \n"),
    writeln(NPossui),
    writeln("        <<Porcentagem figurinhas no álbum (%)>>          \n"),
    P is (Possui / 250) * 100,
    format('~2f', [P]),
    writeln("\n\n"),
    continuar.

contaFigurinhas([], 0, 0).
contaFigurinhas([row(_, true)|T], V, F):- contaFigurinhas(T, V1, F), V is V1 + 1.
contaFigurinhas([row(_, false)|T], V, F):- contaFigurinhas(T, V, F1), F is F1 + 1.